### PR TITLE
Generalize `Upsert`, use it in aggregation.

### PR DESCRIPTION
**Pull request by ryzhyk**
The `Upsert` operator was introduced as part of the implementation of `UpsertHandle` (i.e., an input data source that generates inputs with upsert semantics).  In a nutshell, it takes a vector of `(key, value)` pairs and generates a delta that removes the old value for each `key` and inserts the new value with weight `1`.

Nearly identical logic is used in the implementation of `aggregate`: it computes the new value for each affected key from scratch and then looks up the old aggregate value in its output trace and emits an insertion and a retraction update.  I am currently working on a new operator that implements the SQL `OVER` clause, which also uses upsert for a similar reason.

To avoid duplicate logic, we generalize the `Upsert` operator to work in arbitrary scopes, and break up the implementation of `aggregate` into the actual aggregation logic that computes the new value of the aggregate, and an upsert, which takes care of emitting any necessary insertions and retractions.

### CODE REVIEW

File: src/operator/upsert.rs [hunk 1]
```diff
@@ -1,140 +1,206 @@
...
+    where
+        K: Ord + Clone + Eq + SizeOf + 'static,
+        V: Ord + Clone + Ord + SizeOf + 'static,
+        B::R: ZRingValue + SizeOf,
+        TS: Timestamp + SizeOf,
+        B: Batch<Key = K, Val = V, Time = ()> + 'static,
+    {
+        let circuit = self.circuit();
+
+        // We build the following circuit to implement the upsert semantics.
```
Comment (mbudiu-vmw) [inline src/operator/upsert.rs:49]: can distinct be implemented using upsert too?

**Review by mbudiu-vmw — APPROVED** [review]
I wonder whether this can be generalized to add an arbitrary weight, not just 1.
Then this could be used to implement windows - where upserting with a weight of 0 is removal from the window

File: src/operator/upsert.rs (hunk 1 above)
Comment (ryzhyk) [reply src/operator/upsert.rs:49]: There are two ways to implement `distinct`: (1) assemble output into a trace and use upsert to determine values to retract, (2) compute retractions dynamically at the same time we scan the _input_ trace.  The second option can be more memory-efficient as it doesn't require output trace, and that's what we implement today, hence no need for upsert.

Comment (mbudiu-vmw) [reply src/operator/upsert.rs:49]: but distinct needs to keep the output trace somewhere too, right?

Comment (ryzhyk) [reply src/operator/upsert.rs:49]: It doesn't if you compute the value to retract from the input trace.

Comment (mbudiu-vmw) [reply src/operator/upsert.rs:49]: Is the input trace the integral of the input?
distinct needs to keep the integral around.

Comment (ryzhyk) [reply src/operator/upsert.rs:49]: A trace is the history of the input.  It can be used to compute the integral for any timestamp, current or past.

Comment (mbudiu-vmw) [reply src/operator/upsert.rs:49]: We need a glossary. I think of a trace as Frank's datatype.

Comment (ryzhyk) [reply src/operator/upsert.rs:49]: Yes, it is.

**Pull request merged**